content/en/blog/_posts/2024-04-24-validating-admission-policy-ga/index.md

```diff
@@ -0,0 +1,283 @@
+------
+layout: blog
+title: "Kubernetes 1.30: Validating Admission Policy Is Generally Available"
+slug: validating-admission-policy-ga
+date: 2024-04-24
+---
+
+**Author:** Jiahui Feng (Google)
+
+On behalf of the Kubernetes project, I am excited to announce that ValidatingAdmissionPolicy has reached
+**general availability**
+as part of Kubernetes 1.30 release. If you have not yet read about this new declarative alternative to
+validating admission webhooks, it may be interesting to read our
+[previous post](/blog/2022/12/20/validating-admission-policies-alpha/) about the new feature.
+If you have already heard about Validating Admission Policy and you are eager to try it out, there is no better time to do it now.
+
+Let's have a taste of Validating Admission Policy by replacing a simple webhook. 
+
+## Example admission webhook
+First, let's take a look at an example of a simple webhook. Here is an excerpt from a webhook that
+enforce `runAsNonRoot`, `readOnlyRootFilesystem`, `allowPrivilegeEscalation`, and `privileged` to be set to the least permissive values.
+
+```go
+func verifyDeployment(deploy *appsv1.Deployment) error {
+	var errs []error
+	for i, c := range deploy.Spec.Template.Spec.Containers {
+		if c.Name == "" {
+			return fmt.Errorf("container %d has no name", i)
+		}
+		if c.SecurityContext == nil {
+			errs = append(errs, fmt.Errorf("container %q does not have SecurityContext", c.Name))
+		}
+		if c.SecurityContext.RunAsNonRoot == nil || !*c.SecurityContext.RunAsNonRoot {
+			errs = append(errs, fmt.Errorf("container %q must set RunAsNonRoot to true in its SecurityContext", c.Name))
+		}
+		if c.SecurityContext.ReadOnlyRootFilesystem == nil || !*c.SecurityContext.ReadOnlyRootFilesystem {
+			errs = append(errs, fmt.Errorf("container %q must set ReadOnlyRootFilesystem to true in its SecurityContext", c.Name))
+		}
+		if c.SecurityContext.AllowPrivilegeEscalation != nil && *c.SecurityContext.AllowPrivilegeEscalation {
+			errs = append(errs, fmt.Errorf("container %q must NOT set AllowPrivilegeEscalation to true in its SecurityContext", c.Name))
+		}
+		if c.SecurityContext.Privileged != nil && *c.SecurityContext.Privileged {
+			errs = append(errs, fmt.Errorf("container %q must NOT set Privileged to true in its SecurityContext", c.Name))
+		}
+	}
+	return errors.NewAggregate(errs)
+}
+```
+
```

Comment:
Ideally, put the code from that gist into the Kubernetes website; we'd prefer not to depend on an individual's GitHub account.

You can commit the file into the same directory as the `index.md`.

Reply:
The small problem is that, if I put `webhook.go` in this directory, when one runs `go mod tidy`, whatever `webhook.go` requires, including kubernetes API would be pulled in. Lacking any way to let go toolchain ignore this file, would you recommend that we make it `webhook.go.txt`? Thank you.

Reply:
We don't often run `go mod tidy` against the blog Markdown, so it's likely to be OK to add the file into this repo.
@tengqm - do you have an opinion here?

Reply:
I am okay to embed the code snippet in the blog.

Reply:
OK, @jiahuif - would you be willing to switch away from the link to the gist?

If you need (or would prefer) to achieve something in HTML, like maybe using `<details>` - ask, we may be able to make you a new shortcode.

Reply:
I took the simple approach. I put the source file alongside the page. The CI did not complain so I guess it is fine. Let me know if that looks good to you. Thank you.